Fusaka Mainnet Testing Report (#23)
ethsteel/pm: reports/hardforks/fusaka/Execution-Layer-Mainnet-Testing.md

# Execution Layer Fusaka Mainnet Testing

## Summary

> - Test transactions will be sent to mainnet for each EIP using [EELS's `execute remote` command](https://eest.ethereum.org/main/running_tests/execute/remote/), with the exception of the EIPs in the next bullet point.

Reviewer: ```suggestion
- Test transactions will be sent to mainnet for all Fusaka EIPs that can be tested via [EELS's `execute remote` command](https://eest.ethereum.org/main/running_tests/execute/remote/) (details below).
```

> - Approximately ~[EXECUTE DRY-RUN ETH] Eth gas required for these tests.

Reviewer 2: Do we want to update the value?

```

#### Transaction Hashes

`test_modexp_different_base_lengths[fork_Osaka-state_test-32-bytes-long-base]`: [0x26d0f8b580d735ac74755030f8e4a52ed02814888887e64097c81c7d0f6dd2e0](https://etherscan.io/tx/0x26d0f8b580d735ac74755030f8e4a52ed02814888887e64097c81c7d0f6dd2e0)
`test_modexp_different_base_lengths[fork_Osaka-state_test-33-bytes-long-base]`: [0x71470510279c1006e84fc65c3977983ecdf9658a3aaca0f0e98b90914333e4e7](https://etherscan.io/tx/0x71470510279c1006e84fc65c3977983ecdf9658a3aaca0f0e98b90914333e4e7)
`test_modexp_different_base_lengths[fork_Osaka-state_test-1024-bytes-long-exp]`: [0xca67e591a6f4d4f995ede0927c13e3713d9393660ff9b710fec548bbe6863a51](https://etherscan.io/tx/0xca67e591a6f4d4f995ede0927c13e3713d9393660ff9b710fec548bbe6863a51)
`test_modexp_different_base_lengths[fork_Osaka-state_test-nagydani-1-pow0x10001]`: [0x10d904c6c5d7f3e1f1f527b8871176faac26f4ebc92bdf4a767799fa0e446083](https://etherscan.io/tx/0x10d904c6c5d7f3e1f1f527b8871176faac26f4ebc92bdf4a767799fa0e446083)

> `test_modexp_different_base_lengths[fork_Osaka-state_test-zero-exponent-64bytes]`: [0xf2ae6f732cea8ff243d394790c68150118de7a13edcb3c839a848d1043e63189](https://etherscan.io/tx/0xf2ae6f732cea8ff243d394790c68150118de7a13edcb3c839a848d1043e63189)

Reviewer: ```suggestion
- `test_modexp_different_base_lengths[fork_Osaka-state_test-32-bytes-long-base]`: [0x26d0f8b580d735ac74755030f8e4a52ed02814888887e64097c81c7d0f6dd2e0](https://etherscan.io/tx/0x26d0f8b580d735ac74755030f8e4a52ed02814888887e64097c81c7d0f6dd2e0)
- `test_modexp_different_base_lengths[fork_Osaka-state_test-33-bytes-long-base]`: [0x71470510279c1006e84fc65c3977983ecdf9658a3aaca0f0e98b90914333e4e7](https://etherscan.io/tx/0x71470510279c1006e84fc65c3977983ecdf9658a3aaca0f0e98b90914333e4e7)
- `test_modexp_different_base_lengths[fork_Osaka-state_test-1024-bytes-long-exp]`: [0xca67e591a6f4d4f995ede0927c13e3713d9393660ff9b710fec548bbe6863a51](https://etherscan.io/tx/0xca67e591a6f4d4f995ede0927c13e3713d9393660ff9b710fec548bbe6863a51)
- `test_modexp_different_base_lengths[fork_Osaka-state_test-nagydani-1-pow0x10001]`: [0x10d904c6c5d7f3e1f1f527b8871176faac26f4ebc92bdf4a767799fa0e446083](https://etherscan.io/tx/0x10d904c6c5d7f3e1f1f527b8871176faac26f4ebc92bdf4a767799fa0e446083)
- `test_modexp_different_base_lengths[fork_Osaka-state_test-zero-exponent-64bytes]`: [0xf2ae6f732cea8ff243d394790c68150118de7a13edcb3c839a848d1043e63189](https://etherscan.io/tx/0xf2ae6f732cea8ff243d394790c68150118de7a13edcb3c839a848d1043e63189)
```

.venv/lib/python3.11/site-packages/_pytest/config/__init__.py:812
  /home/marioevz/Development/Eth/execution-specs/.venv/lib/python3.11/site-packages/_pytest/config/__init__.py:812: PytestAssertRewriteWarning: Module already imported so cannot be rewritten; execution_testing.cli.pytest_commands.plugins.shared.execute_fill
    self.import_plugin(arg, consider_entry_points=True)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
------------------------------------------------------- Log file: /home/marioevz/Development/Eth/execution-specs/logs/execute-remote-20251203-221125-main.log --------------------------------------------------------
====================================================================================== 5 passed, 1 warning in 242.22s (0:04:02) ======================================================================================
```

### - EIP-7910: eth_config JSON-RPC Method

#### Description

Run `execute eth-config` after the fork to validate remaining BPO forks.

#### Command

```
uv run execute eth-config --network Mainnet --rpc-endpoint $MAINNET_RPC_ENDPOINT --chain-id=1
```

#### Transaction Hashes

> N/A

Reviewer: ```suggestion
This is a regular call to a node's JSON RPC to request information; it does not require sending a transaction. 
```

#### Outcome

✅ PASS

```
================================================= test session starts ==================================================
platform darwin -- Python 3.13.2, pytest-8.4.2, pluggy-1.6.0
collected 7 items

test_eth_config_current[rpc.srv1.snapshotter.mainnet.ethpandaops.io] .
test_eth_config_current_fork_id[rpc.srv1.snapshotter.mainnet.ethpandaops.io] .
test_eth_config_next[rpc.srv1.snapshotter.mainnet.ethpandaops.io] .
test_eth_config_next_fork_id[rpc.srv1.snapshotter.mainnet.ethpandaops.io] .
test_eth_config_last[rpc.srv1.snapshotter.mainnet.ethpandaops.io] .
test_eth_config_last_fork_id[rpc.srv1.snapshotter.mainnet.ethpandaops.io] .

============================================= 6 passed in 1.33s =============================================
```

> Tested against 5 mainnet RPC endpoints for Geth, Reth, Nethermind, Erigon, and Besu. All returned consistent results.

Author: I don't think we had access to one with ethpandaops, we could check and then update. Thanks!

    },
    "chainId": "0x1",
    "forkId": "0x07c9462e",
    "precompiles": { "..." },
    "systemContracts": { "..." }
  }
}
```

The response shows the next scheduled BPO (Blob Parameter Only) forks, and that Osaka is the current fork:

- **current**: The active Osaka fork with blob target=6, max=9
- **next**: First BPO fork (BPO1) increasing blob parameters to target=10, max=15
- **last**: Final scheduled BPO fork (BPO2) with target=14, max=21

All clients return consistent configurations, confirming correct `eth_config` implementation.

### - EIP-7939: Count leading zeros (CLZ) opcode

#### Description

Execute the opcode with different inputs.

#### Command

<!-- Command used to execute the tests. -->

```
uv run execute remote --fork=Osaka -m mainnet tests/osaka/eip7939_count_leading_zeros/test_eip_mainnet.py --rpc-seed-key $MAINNET_RPC_SEED_KEY --rpc-endpoint $MAINNET_RPC_ENDPOINT --chain-id $MAINNET_CHAIN_ID
```

#### Transaction Hashes

`test_clz_mainnet[fork_Osaka-state_test-clz-8-leading-zeros]`: [0x9ec7c2d94378ce3eaa9c794ce70d2f6f0288918d008bc8de79f167a008cb6437](https://etherscan.io/tx/0x9ec7c2d94378ce3eaa9c794ce70d2f6f0288918d008bc8de79f167a008cb6437)

> `test_clz_mainnet[fork_Osaka-state_test-clz-all-zeros]`: [0x7f9ea62cf3490a0080523c056fb0327ee3de2d9a854e297b8fe86442d5126bb0](https://etherscan.io/tx/0x7f9ea62cf3490a0080523c056fb0327ee3de2d9a854e297b8fe86442d5126bb0)

Reviewer: ```suggestion
- `test_clz_mainnet[fork_Osaka-state_test-clz-8-leading-zeros]`: [0x9ec7c2d94378ce3eaa9c794ce70d2f6f0288918d008bc8de79f167a008cb6437](https://etherscan.io/tx/0x9ec7c2d94378ce3eaa9c794ce70d2f6f0288918d008bc8de79f167a008cb6437)
- `test_clz_mainnet[fork_Osaka-state_test-clz-all-zeros]`: [0x7f9ea62cf3490a0080523c056fb0327ee3de2d9a854e297b8fe86442d5126bb0](https://etherscan.io/tx/0x7f9ea62cf3490a0080523c056fb0327ee3de2d9a854e297b8fe86442d5126bb0)
```

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
------------------------------------------------------- Log file: /home/marioevz/Development/Eth/execution-specs/logs/execute-remote-20251203-221638-main.log --------------------------------------------------------
====================================================================================== 2 passed, 1 warning in 145.33s (0:02:25) ======================================================================================
```

### - EIP-7951: Precompile for secp256r1 Curve Support

#### Description

Execute the precompile with different inputs.

#### Command

<!-- Command used to execute the tests. -->

```
uv run execute remote --fork=Osaka -m mainnet tests/osaka/eip7951_p256verify_precompiles/test_eip_mainnet.py --rpc-seed-key $MAINNET_RPC_SEED_KEY --rpc-endpoint $MAINNET_RPC_ENDPOINT --chain-id $MAINNET_CHAIN_ID
```

#### Transaction Hashes

`test_valid[fork_Osaka-state_test--valid_r1_sig-]`: [0x6911ce5860e375b70c35a5b7d4860a85481e2c4e3db057196cd8089dc7434b92](https://etherscan.io/tx/0x6911ce5860e375b70c35a5b7d4860a85481e2c4e3db057196cd8089dc7434b92)

> `test_invalid[fork_Osaka-state_test--invalid_r1_sig_but_valid_k1_sig-]`: [0x63e3c33f556f13df928c994cbe97e0cb7f5994912fc1c355d128027461d005de](https://etherscan.io/tx/0x63e3c33f556f13df928c994cbe97e0cb7f5994912fc1c355d128027461d005de)

Reviewer: ```suggestion
- `test_valid[fork_Osaka-state_test--valid_r1_sig-]`: [0x6911ce5860e375b70c35a5b7d4860a85481e2c4e3db057196cd8089dc7434b92](https://etherscan.io/tx/0x6911ce5860e375b70c35a5b7d4860a85481e2c4e3db057196cd8089dc7434b92)
- `test_invalid[fork_Osaka-state_test--invalid_r1_sig_but_valid_k1_sig-]`: [0x63e3c33f556f13df928c994cbe97e0cb7f5994912fc1c355d128027461d005de](https://etherscan.io/tx/0x63e3c33f556f13df928c994cbe97e0cb7f5994912fc1c355d128027461d005de)
```

# Execution Layer Fusaka Mainnet Testing

## Summary

- Test transactions will be sent to mainnet for each EIP using [EELS's `execute remote` command](https://eest.ethereum.org/main/running_tests/execute/remote/), with the exception of the EIPs in the next bullet point.

> - Approximately ~0.0001 Eth gas required for these tests.

Reviewer: ```suggestion
- Approximately ~0.0001 ETH gas required for these tests.
```

## EIPs

### Included

- EIP-7823: Set upper bounds for MODEXP
- EIP-7825: Transaction Gas Limit Cap
- EIP-7883: ModExp Gas Cost Increase
- EIP-7910: eth_config JSON-RPC Method
- EIP-7939: Count leading zeros (CLZ) opcode
- EIP-7951: Precompile for secp256r1 Curve Support

### Excluded

- EIP-7594: PeerDAS

Networking related, not testable via EL transactions.

- EIP-7642: eth/69

Networking related, not testable via EL transactions.

- EIP-7892: Blob Parameter Only ('BPO') Hardforks

Requires many blob transactions included in the same block, not easily achievable via `execute` command.

- EIP-7917: Deterministic proposer lookahead

CL only change.

- EIP-7918: Blob base fee bounded by execution cost

Environment dependent, not testable via `execute` command.

- EIP-7934: RLP Execution Block Size Limit

Environment dependent, not testable via `execute` command.

- EIP-7935: Set default gas limit to 60M

Environment dependent, not testable via `execute` command.

## Schedule

- Fork Epoch: 411392 (Dec-03-2025 09:49:11 PM UTC)
- Fork Expected Finalization Epoch: 411394 (Dec-03-2025 10:01:59 PM UTC)

> - Testing will commence at the fork finalization epoch when the chain finalizes successfully without intervention.

Reviewer: Just consistency.
```suggestion
- Fork Epoch: 411392 (Dec-03-2025 09:49:11 PM UTC).
- Fork Expected Finalization Epoch: 411394 (Dec-03-2025 10:01:59 PM UTC).
- Testing will commence at the fork finalization epoch when the chain finalizes successfully without intervention.
```

====================================================================================== 2 passed, 1 warning in 165.43s (0:02:45) ======================================================================================
```

### - EIP-7825: Transaction Gas Limit Cap

#### Description

Send a transaction with the gas limit cap and above.

#### Command

```
uv run execute remote --fork=Osaka -m mainnet tests/osaka/eip7825_transaction_gas_limit_cap/test_eip_mainnet.py --rpc-seed-key $MAINNET_RPC_SEED_KEY --rpc-endpoint $MAINNET_RPC_ENDPOINT --chain-id $MAINNET_CHAIN_ID
```

#### Transaction Hashes

- `test_tx_gas_limit_cap_at_maximum[fork_Osaka-state_test]`: [0x952750833f1dab54ea088dc936f37c6064c0fb3678985d4351f17cca9b0c5ece](https://etherscan.io/tx/0x952750833f1dab54ea088dc936f37c6064c0fb3678985d4351f17cca9b0c5ece)
- `test_tx_gas_limit_cap_exceeded[fork_Osaka-state_test]`: No direct transaction executed (expected):
  - account funding:[0xc5ac11315053776ac903235ef71f97deabe4ae9d5a4bd18b645e97699ecec2f6](https://etherscan.io/tx/0xc5ac11315053776ac903235ef71f97deabe4ae9d5a4bd18b645e97699ecec2f6)

>   - acccount refund after no tx execution: [0x46d32a995a7934d268bc264c7a6e1c15308a73a07cc7c12e50a9ed21a383941b](https://etherscan.io/tx/0x46d32a995a7934d268bc264c7a6e1c15308a73a07cc7c12e50a9ed21a383941b)

Reviewer: Just whitespace.
```suggestion
  - account funding: [0xc5ac11315053776ac903235ef71f97deabe4ae9d5a4bd18b645e97699ecec2f6](https://etherscan.io/tx/0xc5ac11315053776ac903235ef71f97deabe4ae9d5a4bd18b645e97699ecec2f6)
  - acccount refund after no tx execution: [0x46d32a995a7934d268bc264c7a6e1c15308a73a07cc7c12e50a9ed21a383941b](https://etherscan.io/tx/0x46d32a995a7934d268bc264c7a6e1c15308a73a07cc7c12e50a9ed21a383941b)
```
